third_party/txt/src/skia/paragraph_builder_skia.cc

```diff
@@ -0,0 +1,164 @@
+  skia.setStrutStyle(strut_style);
+
+  // General paragraph properties.
+  skia.setTextAlign(static_cast<skt::TextAlign>(txt.text_align));
+  skia.setTextDirection(static_cast<skt::TextDirection>(txt.text_direction));
+  skia.setMaxLines(txt.max_lines);
+  skia.setEllipsis(txt.ellipsis);
+
+  // TODO: break strategy?...
+
+  // This is something flutter expect us to do
+  skia.turnHintingOff();
+
+  return skia;
+}
+
+skt::TextStyle TxtToSkia(const TextStyle& txt) {
+  skt::TextStyle skia;
+
+  skia.setColor(txt.color);
+  skia.setDecoration(static_cast<skt::TextDecoration>(txt.decoration));
+  skia.setDecorationColor(txt.decoration_color);
+  skia.setDecorationStyle(
+      static_cast<skt::TextDecorationStyle>(txt.decoration_style));
+  skia.setDecorationThicknessMultiplier(
+      SkDoubleToScalar(txt.decoration_thickness_multiplier));
+  skia.setFontStyle(MakeSkFontStyle(txt.font_weight, txt.font_style));
+  skia.setTextBaseline(static_cast<skt::TextBaseline>(txt.text_baseline));
+
+  std::vector<SkString> skia_fonts;
+  std::transform(txt.font_families.begin(), txt.font_families.end(),
+                 std::back_inserter(skia_fonts),
+                 [](const std::string& f) { return SkString(f.c_str()); });
+  skia.setFontFamilies(skia_fonts);
+
+  skia.setFontSize(SkDoubleToScalar(txt.font_size));
+  skia.setLetterSpacing(SkDoubleToScalar(txt.letter_spacing));
+  skia.setWordSpacing(SkDoubleToScalar(txt.word_spacing));
+  skia.setHeight(SkDoubleToScalar(txt.height));
+
+  skia.setLocale(SkString(txt.locale.c_str()));
+  if (txt.has_background) {
+    skia.setBackgroundColor(txt.background);
+  }
+  if (txt.has_foreground) {
+    skia.setForegroundColor(txt.foreground);
+  }
+
+  skia.resetShadows();
+  for (const txt::TextShadow& txt_shadow : txt.text_shadows) {
+    skt::TextShadow shadow;
+    shadow.fOffset = txt_shadow.offset;
+    shadow.fBlurRadius = txt_shadow.blur_radius;
+    shadow.fColor = txt_shadow.color;
+    skia.addShadow(shadow);
+  }
+
+  return skia;
+}
+
+}  // anonymous namespace
+
+ParagraphBuilderSkia::ParagraphBuilderSkia(
+    const ParagraphStyle& style,
+    std::shared_ptr<FontCollection> font_collection)
+    : builder_(skt::ParagraphBuilder::make(
+          TxtToSkia(style),
+          font_collection->CreateSktFontCollection())),
+      base_style_(style.GetTextStyle()) {}
+
+ParagraphBuilderSkia::~ParagraphBuilderSkia() = default;
+
+void ParagraphBuilderSkia::PushStyle(const TextStyle& style) {
+  builder_->pushStyle(TxtToSkia(style));
+  txt_style_stack_.push(style);
+}
+
+void ParagraphBuilderSkia::Pop() {
+  builder_->pop();
+  txt_style_stack_.pop();
+}
+
+const TextStyle& ParagraphBuilderSkia::PeekStyle() {
+  return txt_style_stack_.empty() ? base_style_ : txt_style_stack_.top();
+}
+
+void ParagraphBuilderSkia::AddText(const std::u16string& text) {
+  builder_->addText(text);
+}
+
+void ParagraphBuilderSkia::AddPlaceholder(PlaceholderRun& span) {
+  // TODO: implement
```

Comment:
Should we have an assert(false) or log statement in the meantime?

Reply:
done